### PR TITLE
okhttp: Reset stream when receiving window is negative  Resolves #215

### LINKED ISSUE #215: OkHttpClientStream.window is only written never read
@ejona86 

Should be removed

### PR DESCRIPTION
**Pull request by madongfly**
Resets the stream when the receiving window size goes negative.

### CODE REVIEW

File: core/src/main/java/io/grpc/transport/AbstractClientStream.java [hunk 1]
```diff
@@ -98,7 +98,7 @@ protected void inboundTransportError(Status errorStatus) {
     }
     // For transport errors we immediately report status to the application layer
     // and do not wait for additional payloads.
-    transportReportStatus(errorStatus, false, new Metadata.Trailers());
+    transportReportStatus(errorStatus, true, new Metadata.Trailers());
```
Comment (ejona86) [inline core/src/main/java/io/grpc/transport/AbstractClientStream.java]: Why changing this to true? It looks like it was correct before.

File: okhttp/src/main/java/io/grpc/transport/okhttp/OkHttpClientStream.java [hunk 2]
```diff
@@ -103,13 +103,23 @@ public void transportHeadersReceived(List<Header> headers, boolean endOfStream)
...
    */
   public void transportDataReceived(okio.Buffer frame, boolean endOfStream) {
     synchronized (lock) {
       long length = frame.size();
       window -= length;
+      if (window < 0) {
+        frameWriter.rstStream(id(), ErrorCode.FLOW_CONTROL_ERROR);
+        Status status = Status.INTERNAL.withDescription(
+            "Received data size exceeded our receiving window size");
+        this.inboundTransportError(status);
```
Comment (ejona86) [inline okhttp/src/main/java/io/grpc/transport/okhttp/OkHttpClientStream.java]: Won't this be done by finishStream()?

**Comment by ejona86** [comment]
Please give commits descriptions that are useful at a glance. You can include the issue number that it fixes, but that shouldn't be part of the short one-line commit description.

**Comment by ejona86** [comment]
This would be a better commit message:
```
okhttp: Reset stream when receiving window is negative

Resolves #215
```

File: core/src/main/java/io/grpc/transport/AbstractClientStream.java (hunk 1 above)
Comment (madongfly) [inline core/src/main/java/io/grpc/transport/AbstractClientStream.java]: According to the comment "For transport errors we immediately report status to the application layer and do not wait for additional payloads.", I thought this was a typo?

File: okhttp/src/main/java/io/grpc/transport/okhttp/OkHttpClientStream.java (hunk 2 above)
Comment (madongfly) [inline okhttp/src/main/java/io/grpc/transport/okhttp/OkHttpClientStream.java]: Good catch, done.

File: core/src/main/java/io/grpc/transport/AbstractClientStream.java (hunk 1 above)
Comment (ejona86) [inline core/src/main/java/io/grpc/transport/AbstractClientStream.java]: The comment isn't really wrong, but is misleading. We should let the error queue behind any _previously_ received data. The only case where it makes a lot of sense to me to jump the queue (and use ```true```) is for cancellation.

**Comment by ejona86** [comment]
@madongfly, LGTM

**Comment by ejona86** [comment]
Don't forget to improve the commit message before bringing into master.

**Pull request merged**